### DISCUSSION TITLE
Single channel Tiffs in Vitessce

### DISCUSSION THREAD

**Discussion by lee-t** [vitessce/vitessce-python#188] [Ideas]
I'm trying to get Merfish images from https://datadryad.org/stash/dataset/doi:10.5061/dryad.jm63xsjb2 which are two single-channel tiffs with 8 z-layers, one dapi stain and another for Na+/K+-ATPase.  I would like to combine these images with either SpatialExperiment or anndata for downstream analysis but I'm having issues with these data formats; they tend to break on image data when using the supplied wrappers and I don't think there is any support for these images in vitessce anyways if I tried to write my own wrapper. Is there a way to import single channel tiffs into vitessce or viv? This way I can separate the cell-label, cell-boundary data, and molecule data and plot that spatially over the tiffs for the stain. Otherwise, I'll need to find a way to get this data into ome-tiff or ome-ngff so I can visualize the data in Vitessce.

**Comment by keller-mark** [comment]
We are working on this feature in the underlying image rendering library Viv https://github.com/hms-dbmi/viv/issues/609 but currently only OME-TIFF and OME-NGFF ([v3](https://github.com/vitessce/vitessce/blob/main/docs/docs/data-troubleshooting.md#rasterome-zarr-troubleshooting)) are supported.

**Reply by lee-t** [reply]
Any suggestions then on how to go from single-channel tiffs into OME-TIFF and OME-NGFF?

**Reply by keller-mark** [reply]
## To go from multiple single-channel tiffs to one OME-NGFF:

```sh
pip install tifffile numpy zarr numcodecs ome-zarr==0.2.1
```

```py
import zarr
import numpy as np
from tifffile import imread
from ome_zarr import writer

c1 = imread("my_image_c1.tif")
c2 = imread("my_image_c2.tif")
my_image = np.stack((c1, c2)) # if using axes = "czyx" in write_image, then the numpy array axes should also be orderd czyx

z_root = zarr.open_group("my_image.zarr", mode = "w")

default_window = {
    "start": 0,
    "min": 0,
    "max": 65_535, # may need to change depending on the numpy dtype of the my_image array
    "end": 65_535 # may need to change depending on the numpy dtype of the my_image array
}

writer.write_image(
    image = my_image,
    group = z_root,
    axes = "czyx",
    omero = {
        "name": "My image",
        "version": "0.3",
        "rdefs": {},
        "channels": [
            {
                "label": f"Channel {i}", # may want to use a different name for each channel
                "color": "FFFFFF", # may want to use a different color for each channel
                "window": default_window
            } for i in range(my_image.shape[0])
        ]
    },
    chunks = (1, 1, 256, 256),
)
```

## To go from multiple single-channel tiffs to one OME-TIFF:

You can use FIJI to merge multiple single-channel tiffs into a single multi-channel tiff.
- From the GUI, see Image -> Hyperstacks
- From a script, see [t_to_z.js](https://github.com/keller-mark/vitessce-baysor-processing/blob/main/src/t_to_z.js) and corresponding [Snakefile](https://github.com/keller-mark/vitessce-baysor-processing/blob/main/Snakefile#L43) (lines 10 and 43)

Then use [bftools](https://docs.openmicroscopy.org/bio-formats/5.7.1/users/comlinetools/index.html) (which comes with the `bfconvert` command line tool) to convert from tiff to OME-TIFF.

```sh
bfconvert -tilex 512 -tiley 512 -pyramid-resolutions 6 -pyramid-scale 2  -compression LZW my_input.tiff my_output.ome.tiff
```

## To add channel names to OME-TIFF

If you want to add channel names, you will need to set them via OME-XML which you can get/set within an OME-TIFF using `tiffcomment` (which also comes with bftools)

The channel names come from the “Name” property of the Channel nodes in the OME-XML. See OME -> Image -> Pixels -> Channel.Name.

Then the [bftools/tiffcomment](https://docs.openmicroscopy.org/bio-formats/5.9.1/users/comlinetools/index.html) command line tool can be used to get and set OME-XML metadata in an OME-TIFF.

```sh
tiffcomment my_image.ome.tiff > in.ome.xml
python add_channel_names.py -i in.ome.xml -o out.ome.xml
tiffcomment -set 'out.ome.xml' my_image.ome.tiff
```

In [add_channel_names.py](https://github.com/keller-mark/vitessce-mesmer-processing/blob/main/src/add_channel_names.py) I used the beautifulsoup package to generate XML - it was before I knew about the [ome-types](https://ome-types.readthedocs.io/en/latest/ome_types.html#ome_types.schema.to_xml) package which has functions for writing OME-XML.

**Reply by keller-mark** [reply]
I recommend opening OME-TIFF files using FIJI and inspecting the OME-XML for debugging metadata such as channel names, etc.

**Reply by lee-t** [reply]
Thanks a bunch for that Mark! that gives us certainly some directions to follow up on. I also recently noticed we do not have scalefactors or tiling but if these tags are optional then I'm definitely going to give your code a try.

**Comment by cgohlke** [comment]
> a way to get this data into ome-tiff 

```Python
import tifffile
from scipy.ndimage import zoom

channels = ['dapi_stack.tif', 'membrane_stack.tif']

czyx_stack = tifffile.imread(channels)

levels = 4
tile = (256, 256)
compression = 'adobe_deflate'

with tifffile.TiffWriter('czyx_stack.ome.tiff', bigtiff=True) as tif:
    tif.write(
        czyx_stack,
        subifds=levels,
        tile=tile,
        compression=compression,
        metadata={'axes': 'czyx', 'Channel': {'Name': channels}},
    )
    for level in range(levels):
        czyx_stack = zoom(
            czyx_stack, (1, 1, 0.5, 0.5), order=1, prefilter=False
        )
        tif.write(
            czyx_stack,
            subfiletype=1,
            tile=tile,
            compression=compression,
        )
```

**Reply by lee-t** [reply]
Thank you for looking at the data! This proved easier than using FIJI ironically.

**Comment by lee-t** [comment]
Thanks to @cgohlke and @keller-mark   I end up with this [configuration ](http://vitessce.io/#?edit=false&url=data%3A%2C%250A%257B%2522version%2522%253A%2520%25221.0.7%2522%252C%2520%2522name%2522%253A%2520%2522ileum%2520tiff%2522%252C%2520%2522description%2522%253A%2520%2522converted%2520single%2520channel%2520tiffs%2520to%2520ometiff%2522%252C%2520%2522datasets%2522%253A%2520%255B%257B%2522uid%2522%253A%2520%2522A%2522%252C%2520%2522name%2522%253A%2520%2522jeff%2522%252C%2520%2522files%2522%253A%2520%255B%257B%2522type%2522%253A%2520%2522raster%2522%252C%2520%2522fileType%2522%253A%2520%2522raster.json%2522%252C%2520%2522options%2522%253A%2520%257B%2522schemaVersion%2522%253A%2520%25220.0.2%2522%252C%2520%2522images%2522%253A%2520%255B%257B%2522name%2522%253A%2520%2522ileum%2522%252C%2520%2522type%2522%253A%2520%2522ome-tiff%2522%252C%2520%2522url%2522%253A%2520%2522https%253A%252F%252Fhmsrc-ccb-data1.s3.amazonaws.com%252Fileum-spatial%252Fileum_stack.ome.tiff%2522%252C%2520%2522metadata%2522%253A%2520%257B%2522isBitmask%2522%253A%2520false%257D%257D%255D%257D%257D%255D%257D%255D%252C%2520%2522coordinationSpace%2522%253A%2520%257B%2522dataset%2522%253A%2520%257B%2522A%2522%253A%2520%2522A%2522%257D%257D%252C%2520%2522layout%2522%253A%2520%255B%257B%2522component%2522%253A%2520%2522spatial%2522%252C%2520%2522coordinationScopes%2522%253A%2520%257B%2522dataset%2522%253A%2520%2522A%2522%257D%252C%2520%2522x%2522%253A%25200.0%252C%2520%2522y%2522%253A%25200%252C%2520%2522w%2522%253A%25206.0%252C%2520%2522h%2522%253A%252012%257D%252C%2520%257B%2522component%2522%253A%2520%2522status%2522%252C%2520%2522coordinationScopes%2522%253A%2520%257B%2522dataset%2522%253A%2520%2522A%2522%257D%252C%2520%2522x%2522%253A%25200%252C%2520%2522y%2522%253A%25200%252C%2520%2522w%2522%253A%25201%252C%2520%2522h%2522%253A%25201%257D%252C%2520%257B%2522component%2522%253A%2520%2522layerController%2522%252C%2520%2522coordinationScopes%2522%253A%2520%257B%2522dataset%2522%253A%2520%2522A%2522%257D%252C%2520%2522x%2522%253A%25206.0%252C%2520%2522y%2522%253A%25200%252C%2520%2522w%2522%253A%25206.0%252C%2520%2522h%2522%253A%252012%252C%2520%2522props%2522%253A%2520%257B%2522disableChannelsIfRgbDetected%2522%253A%2520true%257D%257D%255D%252C%2520%2522initStrategy%2522%253A%2520%2522auto%2522%257D)for vitessce. As a follow up, is there a way to add in molecule or cell set data from another dataset? 
If I configure a viewer in vitessce  
``` python
vc = VitessceConfig(name='merfish', description='MERFISH measurements in the mouse ileum')
dataset = vc.add_dataset(name='ileum').add_object(AnnDataWrapper(gdata, 
    cell_set_obs=["leiden_final"], cell_set_obs_names=["Leiden"], spatial_centroid_obsm="spatial" , expression_matrix="X" ))
dataset2 = vc.add_dataset(name='jeff').add_object(
    OmeTiffWrapper(img_path="czyx_stack.ome.tiff", name='ileum'),)
spatial = vc.add_view(cm.SPATIAL, dataset=dataset)
tiff = vc.add_view(cm.SPATIAL, dataset=dataset2)
```
I get a vitessce instance with 2 coordinated spatial windows, one with the tiff and a corresponding plot with centroids and cell labels. Is there a way to overlay `dataset` onto `dataset2` in vitessce? doesn't seem to work that way unless I'm missing something. Otherwise the next step I guess is to annotate the ome-tiff with cell labels or covert to ome-ngff while annotating labels such that the image data and cell set metadata are the same source dataset.

**Reply by lee-t** [reply]
Nevermind, I read that [add_dataset()](https://vitessce.github.io/vitessce-python/api_config.html#vitessce.config.VitessceConfig.add_dataset) has an argument that accepts a list of objects. This will work:
```python
vc = VitessceConfig(name='merfish', description='MERFISH measurements in the mouse ileum')
dataset = vc.add_dataset(name='ileum', objs= [AnnDataWrapper(gdata, cell_set_obs=["leiden_final"], cell_set_obs_names=["Leiden"], mappings_obsm=["X_umap", "X_pca"], mappings_obsm_names=["UMAP", "PCA"], spatial_centroid_obsm="spatial" , expression_matrix="X"), 
    OmeTiffWrapper(img_path="czyx_stack.ome.tiff", name='ileum' )])
```

**Reply by keller-mark** [reply]
add_object also returns the dataset instance so you can also chain the calls like

```
dataset = (vc.add_dataset(name='ileum')
  .add_object(AnnDataWrapper(gdata, cell_set_obs=["leiden_final"], cell_set_obs_names=["Leiden"], spatial_centroid_obsm="spatial" , expression_matrix="X" ))
  .add_object(OmeTiffWrapper(img_path="czyx_stack.ome.tiff", name='ileum'))
)
